### PR TITLE
Dialog fragment implementation

### PR DESCRIPTION
**Pull request by juanii000**
Creation of dialog fragment to set the number of slots on the parking.

Pixel 4
dialogFragmentTest
Nexus S
nexuss

image

### CODE REVIEW

File: app/src/main/java/com/example/parkingsystem_kotlin/mvp/contract/ParkingContract.kt [hunk 1]
```diff
@@ -3,13 +3,15 @@ package com.example.parkingsystem_kotlin.mvp.contract
 interface ParkingContract {
     interface Presenter{
         fun onParkingSizeButtonPressed()
+        fun onParkingSizeSet(parkingSize: String)
     }
 
     interface Model{
-        fun getParkingSize(): Int
+        var parkingSize: Int;
```
Comment (SartiMau) [inline app/src/main/java/com/example/parkingsystem_kotlin/mvp/contract/ParkingContract.kt]: remember that the ; is not necessary in kotlin so please remove it. On the other hand, do not put variables on the contacts, just keet a method to keep the encapsulation

File: app/src/main/java/com/example/parkingsystem_kotlin/mvp/model/ParkingModel.kt [hunk 2]
```diff
@@ -1,11 +1,6 @@
...
-import com.example.parkingsystem_kotlin.util.Constants.PARKING_SIZE
+import com.example.parkingsystem_kotlin.util.Constants
 
-class ParkingModel: ParkingContract.Model {
-
-    private val parkingSize: Int = PARKING_SIZE
-
-    override fun getParkingSize(): Int = parkingSize
-}
+class ParkingModel(override var parkingSize: Int = Constants.PARKING_SIZE) : ParkingContract.Model
```
Comment (SartiMau) [inline app/src/main/java/com/example/parkingsystem_kotlin/mvp/model/ParkingModel.kt]: please undo this change to keep the encapsulation

File: app/src/main/java/com/example/parkingsystem_kotlin/mvp/presenter/ParkingSizeDialogPresenter.kt [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+import com.example.parkingsystem_kotlin.mvp.contract.ParkingSizeDialogContract
+
+class ParkingSizeDialogPresenter(
+    private val view: ParkingSizeDialogContract.ParkingSizeView,
+) : ParkingSizeDialogContract.ParkingSizePresenter {
+
+    override fun setParkingSize(inputListener: SpacesParkingDialogListener) {
+        val parkingSize = view.getNewParkingSize()
+        try {
+            if (parkingSize.isNotEmpty() && parkingSize.toInt() < 10000) {
```
Comment (SartiMau) [inline app/src/main/java/com/example/parkingsystem_kotlin/mvp/presenter/ParkingSizeDialogPresenter.kt]: do not hardcode please, use a constant

File: app/src/main/java/com/example/parkingsystem_kotlin/mvp/presenter/ParkingSizeDialogPresenter.kt [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+                inputListener.setParkingSize(parkingSize)
+                view.closeDialog()
+            } else {
+                view.showToastEmptyInput()
+            }
+        }catch (e: NumberFormatException){
```
Comment (SartiMau) [inline app/src/main/java/com/example/parkingsystem_kotlin/mvp/presenter/ParkingSizeDialogPresenter.kt]: I think that the try/catch is not necessary, to avoid this you can set on the dialog XML the input type as number so the user will only be able to put numbers

File: app/src/main/java/com/example/parkingsystem_kotlin/mvp/view/ParkingSizeDialogView.kt [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+    override fun showToastException() {
+        context?.let {
+            it.toast(it.getString(R.string.toast_input_invalid_number))
+        }
+    }
+
+    override fun closeDialog() {
+        fragment?.let {
+            val dialog: DialogFragment = it as DialogFragment
```
Comment (SartiMau) [inline app/src/main/java/com/example/parkingsystem_kotlin/mvp/view/ParkingSizeDialogView.kt]: you can improves this by 
`(it as DialogFragment).dismiss()`

File: app/src/main/java/com/example/parkingsystem_kotlin/mvp/view/ParkingView.kt [hunk 6]
```diff
@@ -2,11 +2,19 @@ package com.example.parkingsystem_kotlin.mvp.view
...
+import androidx.fragment.app.DialogFragment
+import com.example.parkingsystem_kotlin.fragment.ParkingSizeDialogFragment
 import com.example.parkingsystem_kotlin.mvp.contract.ParkingContract
 import com.example.parkingsystem_kotlin.mvp.view.base.ActivityView
+import com.example.parkingsystem_kotlin.util.Constants
 
 class ParkingView(activity: Activity): ActivityView(activity), ParkingContract.View {
 
+    override fun showParkingSizeDialogFragment() {
+        val dialogFragment: DialogFragment = ParkingSizeDialogFragment();
```
Comment (SartiMau) [inline app/src/main/java/com/example/parkingsystem_kotlin/mvp/view/ParkingView.kt]: do not use ; please

File: app/src/main/res/layout/dialog_fragment_parking_size.xml [hunk 7]
```diff
@@ -0,0 +1,54 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    android:layout_margin="@dimen/margin_25dp">
```
Comment (SartiMau) [inline app/src/main/res/layout/dialog_fragment_parking_size.xml]: please do not put numbers in the dimens key. you can change it by margin_small, margin_medium or general_margin for example

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/res/layout/activity_main.xml [hunk 8]
```diff
@@ -16,9 +16,10 @@
 
     <Button
         android:id="@+id/button_main_parking_size"
-        style="@style/ButtonConfigureParkingSize"
+        style="@style/Button"
         android:layout_width="match_parent"
         android:layout_height="wrap_content"
+        android:text="@string/button_configure_parking_size"
         app:layout_constraintTop_toBottomOf="@+id/main_toolbar"/>
```
Comment (mauroLonghi) [inline app/src/main/res/layout/activity_main.xml]: not use "+" to contraint the views. you only need to call :  id/main_toolbar

File: app/src/main/res/values/dimens.xml [hunk 9]
```diff
@@ -1,4 +1,5 @@
 <?xml version="1.0" encoding="utf-8"?>
 <resources>
-    <dimen name="margin_main_buttons">25dp</dimen>
+    <dimen name="margin_25dp">25dp</dimen>
+    <dimen name="text_title">35sp</dimen>
```
Comment (mauroLonghi) [inline app/src/main/res/values/dimens.xml]: In both cases, not allignment to the mobile code style. pls rename it

File: app/src/main/res/values/strings.xml [hunk 10]
```diff
@@ -1,4 +1,10 @@
 <resources>
     <string name="app_name">ParkingSystem-Kotlin</string>
     <string name="button_configure_parking_size">configure parking size</string>
+    <string name="button_set_parking_size">set parking size</string>
+    <string name="button_exit_parking_size">exit parking size</string>
+    <string name="title_parking_size">Parking Size</string>
+    <string name="input_hint_insert_number">Insert a number</string>
+    <string name="toast_empty_input_msg">Please, complete the field</string>
+    <string name="toast_input_invalid_number">The number can\'t be greater than 9999</string>
```
Comment (mauroLonghi) [inline app/src/main/res/values/strings.xml]: use the code style guidelines

File: app/src/main/res/values/styles.xml [hunk 11]
```diff
@@ -6,14 +6,15 @@
         <item name="android:title">@string/app_name</item>
         <item name="android:titleTextColor">@color/white</item>
     </style>
-    
-    <style name="ButtonMainPageStyle">
-        <item name="backgroundTint">@color/teal_700</item>
-        <item name="android:layout_margin">@dimen/margin_main_buttons</item>
+
+    <style name="TitleText">
```
Comment (mauroLonghi) [inline app/src/main/res/values/styles.xml]: rename it

File: app/src/main/java/com/example/parkingsystem_kotlin/mvp/presenter/ParkingSizeDialogPresenter.kt (hunk 4 above)
Comment (juanii000) [reply app/src/main/java/com/example/parkingsystem_kotlin/mvp/presenter/ParkingSizeDialogPresenter.kt]: Im using this try catch because when I use very big numbers it brokes the .toInt, maybe there is a way to solve that in another way, let me know :P

File: app/src/main/java/com/example/parkingsystem_kotlin/mvp/model/ParkingModel.kt [hunk 12]
```diff
@@ -1,11 +1,13 @@
 package com.example.parkingsystem_kotlin.mvp.model
 
 import com.example.parkingsystem_kotlin.mvp.contract.ParkingContract
-import com.example.parkingsystem_kotlin.util.Constants.PARKING_SIZE
+import com.example.parkingsystem_kotlin.util.Constants
 
-class ParkingModel: ParkingContract.Model {
+class ParkingModel(private var parkingSize: Int = Constants.PARKING_SIZE) : ParkingContract.Model {
```
Comment (SartiMau) [inline app/src/main/java/com/example/parkingsystem_kotlin/mvp/model/ParkingModel.kt]: you are not setting this in the constructor in the activity, you are just using it after the creation so please, put it as a variable outside the constructor

File: app/src/main/res/values/styles.xml [hunk 13]
```diff
@@ -6,14 +6,15 @@
...
+        <item name="android:textSize">@dimen/size_all_title_xbig</item>
+        <item name="android:textStyle">bold</item>
     </style>
-    
-    <style name="ButtonConfigureParkingSize" parent="ButtonMainPageStyle">
-        <item name="android:text">@string/button_configure_parking_size</item>
+
+    <style name="Button">
+        <item name="backgroundTint">@color/teal_700</item>
+        <item name="android:textSize">12sp</item>
```
Comment (SartiMau) [inline app/src/main/res/values/styles.xml]: use dimens pls, no hardcode

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/res/values/styles.xml (hunk 11 above)
Comment (juanii000) [reply app/src/main/res/values/styles.xml]: done

File: app/src/main/res/values/strings.xml (hunk 10 above)
Comment (juanii000) [reply app/src/main/res/values/strings.xml]: done

File: app/src/main/res/values/dimens.xml (hunk 9 above)
Comment (juanii000) [reply app/src/main/res/values/dimens.xml]: done

File: app/src/main/res/layout/activity_main.xml (hunk 8 above)
Comment (juanii000) [reply app/src/main/res/layout/activity_main.xml]: done

File: app/src/main/res/layout/dialog_fragment_parking_size.xml (hunk 7 above)
Comment (juanii000) [reply app/src/main/res/layout/dialog_fragment_parking_size.xml]: done

File: app/src/main/java/com/example/parkingsystem_kotlin/mvp/view/ParkingView.kt (hunk 6 above)
Comment (juanii000) [reply app/src/main/java/com/example/parkingsystem_kotlin/mvp/view/ParkingView.kt]: done

File: app/src/main/java/com/example/parkingsystem_kotlin/mvp/view/ParkingSizeDialogView.kt (hunk 5 above)
Comment (juanii000) [reply app/src/main/java/com/example/parkingsystem_kotlin/mvp/view/ParkingSizeDialogView.kt]: done

File: app/src/main/java/com/example/parkingsystem_kotlin/mvp/presenter/ParkingSizeDialogPresenter.kt (hunk 4 above)
Comment (juanii000) [reply app/src/main/java/com/example/parkingsystem_kotlin/mvp/presenter/ParkingSizeDialogPresenter.kt]: done

File: app/src/main/java/com/example/parkingsystem_kotlin/mvp/presenter/ParkingSizeDialogPresenter.kt (hunk 3 above)
Comment (juanii000) [reply app/src/main/java/com/example/parkingsystem_kotlin/mvp/presenter/ParkingSizeDialogPresenter.kt]: done

File: app/src/main/java/com/example/parkingsystem_kotlin/mvp/model/ParkingModel.kt (hunk 2 above)
Comment (juanii000) [reply app/src/main/java/com/example/parkingsystem_kotlin/mvp/model/ParkingModel.kt]: done

File: app/src/main/java/com/example/parkingsystem_kotlin/mvp/contract/ParkingContract.kt (hunk 1 above)
Comment (juanii000) [reply app/src/main/java/com/example/parkingsystem_kotlin/mvp/contract/ParkingContract.kt]: done

File: app/src/main/java/com/example/parkingsystem_kotlin/fragment/ParkingSizeDialogFragment.kt [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        presenter = ParkingSizeDialogPresenter(ParkingSizeDialogView(this, binding))
+    }
+
+    private fun setOnClickListeners() {
+        binding.buttonSetParkingSize.setOnClickListener { presenter.setParkingSize(inputListener) }
+        binding.buttonExitParkingSize.setOnClickListener { presenter.exitDialog() }
```
Comment (FranciscoGaraventa) [inline app/src/main/java/com/example/parkingsystem_kotlin/fragment/ParkingSizeDialogFragment.kt]: In this case you could use with(binding){ ... }

File: app/src/main/res/values/strings.xml (hunk 10 above)
Comment (mauroLonghi) [reply app/src/main/res/values/strings.xml]: mmm it still wrong , try to rename

File: app/src/main/res/values/styles.xml [hunk 15]
```diff
@@ -6,14 +6,15 @@
...
+
+    <style name="TextStyleAllAppTitle">
+        <item name="android:textSize">@dimen/size_all_title_xbig</item>
+        <item name="android:textStyle">bold</item>
     </style>
-    
-    <style name="ButtonConfigureParkingSize" parent="ButtonMainPageStyle">
-        <item name="android:text">@string/button_configure_parking_size</item>
+
+    <style name="Button">
```
Comment (mauroLonghi) [inline app/src/main/res/values/styles.xml]: rename it pls

Comment (juanii000) [reply app/src/main/res/values/styles.xml]: Done.

File: app/src/main/java/com/example/parkingsystem_kotlin/fragment/ParkingSizeDialogFragment.kt (hunk 14 above)
Comment (juanii000) [reply app/src/main/java/com/example/parkingsystem_kotlin/fragment/ParkingSizeDialogFragment.kt]: Done.

File: app/src/main/res/values/styles.xml (hunk 13 above)
Comment (juanii000) [reply app/src/main/res/values/styles.xml]: Done.

File: app/src/main/java/com/example/parkingsystem_kotlin/mvp/model/ParkingModel.kt (hunk 12 above)
Comment (juanii000) [reply app/src/main/java/com/example/parkingsystem_kotlin/mvp/model/ParkingModel.kt]: Done.

File: app/src/main/res/values/strings.xml (hunk 10 above)
Comment (juanii000) [reply app/src/main/res/values/strings.xml]: Done.

File: app/src/main/java/com/example/parkingsystem_kotlin/util/ExtensionFunctions.kt [hunk 16]
```diff
@@ -0,0 +1,7 @@
+package com.example.parkingsystem_kotlin.util
+
+import android.content.Context
+import android.widget.Toast
+
+fun Context.toast(message: String) =
+    Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
```
Comment (mauroLonghi) [inline app/src/main/java/com/example/parkingsystem_kotlin/util/ExtensionFunctions.kt]: nice !

File: app/src/main/java/com/example/parkingsystem_kotlin/mvp/presenter/ParkingSizeDialogPresenter.kt [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+import com.example.parkingsystem_kotlin.mvp.contract.ParkingSizeDialogContract
+import com.example.parkingsystem_kotlin.util.Constants
+
+class ParkingSizeDialogPresenter(
+    private val view: ParkingSizeDialogContract.ParkingSizeView,
+) : ParkingSizeDialogContract.ParkingSizePresenter {
+
+    override fun setParkingSize(inputListener: SpacesParkingDialogListener) {
+        val parkingSize = view.getNewParkingSize()
+        if (parkingSize.isNotEmpty() && parkingSize.toLong() < Constants.PARKING_MAX_SIZE) {
```
Comment (mauroLonghi) [inline app/src/main/java/com/example/parkingsystem_kotlin/mvp/presenter/ParkingSizeDialogPresenter.kt]: You could define this constant in the companion object

**Review by SartiMau — APPROVED** [review]

**Review by FranciscoGaraventa — APPROVED** [review]
LG

File: app/src/main/java/com/example/parkingsystem_kotlin/mvp/presenter/ParkingSizeDialogPresenter.kt (hunk 17 above)
Comment (juanii000) [reply app/src/main/java/com/example/parkingsystem_kotlin/mvp/presenter/ParkingSizeDialogPresenter.kt]: Done !

**Review by mauroLonghi — APPROVED** [review]

File: app/src/main/java/com/example/parkingsystem_kotlin/mvp/view/ParkingView.kt [hunk 18]
```diff
@@ -2,11 +2,19 @@ package com.example.parkingsystem_kotlin.mvp.view
...
+import com.example.parkingsystem_kotlin.fragment.ParkingSizeDialogFragment
 import com.example.parkingsystem_kotlin.mvp.contract.ParkingContract
 import com.example.parkingsystem_kotlin.mvp.view.base.ActivityView
+import com.example.parkingsystem_kotlin.util.Constants
 
 class ParkingView(activity: Activity): ActivityView(activity), ParkingContract.View {
 
+    override fun showParkingSizeDialogFragment() {
+        val dialogFragment: DialogFragment = ParkingSizeDialogFragment()
+        dialogFragment.show(fragmentManager, Constants.TAG_PARKING_SIZE_DIALOG)
```
Comment (nicolbidone) [inline app/src/main/java/com/example/parkingsystem_kotlin/mvp/view/ParkingView.kt]: there's no need to create an extra var yo can call  ParkingSizeDialogFragment().show(...

File: app/src/main/java/com/example/parkingsystem_kotlin/util/ExtensionFunctions.kt (hunk 16 above)
Comment (nicolbidone) [inline app/src/main/java/com/example/parkingsystem_kotlin/util/ExtensionFunctions.kt]: this can fit on a single line

File: app/src/main/res/layout/dialog_fragment_parking_size.xml [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+    <EditText
+        android:id="@+id/number_input_dialog_fragment_parking_size"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:layout_margin="@dimen/margin_all_general_normal"
+        android:autofillHints="no"
+        android:hint="@string/input_dialog_parking_space_hint_insert_number"
+        android:inputType="number"
+        app:layout_constraintLeft_toLeftOf="parent"
+        app:layout_constraintRight_toRightOf="parent"
```
Comment (nicolbidone) [inline app/src/main/res/layout/dialog_fragment_parking_size.xml:28]: if you have width match parent horizontal constraints are useless

File: app/src/main/res/layout/dialog_fragment_parking_size.xml [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+        app:layout_constraintTop_toTopOf="parent" />
+
+    <EditText
+        android:id="@+id/number_input_dialog_fragment_parking_size"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:layout_margin="@dimen/margin_all_general_normal"
+        android:autofillHints="no"
```
Comment (nicolbidone) [inline app/src/main/res/layout/dialog_fragment_parking_size.xml:24]: nice 👍🏻

File: app/src/main/java/com/example/parkingsystem_kotlin/util/ExtensionFunctions.kt (hunk 16 above)
Comment (juanii000) [reply app/src/main/java/com/example/parkingsystem_kotlin/util/ExtensionFunctions.kt]: The Android Studio is a very strict Indentator :p 
image

File: app/src/main/res/layout/dialog_fragment_parking_size.xml (hunk 19 above)
Comment (juanii000) [reply app/src/main/res/layout/dialog_fragment_parking_size.xml:28]: Nice tip

Comment (juanii000) [reply app/src/main/res/layout/dialog_fragment_parking_size.xml:28]: But I think I need at least one for this occasion

File: app/src/main/java/com/example/parkingsystem_kotlin/mvp/view/ParkingView.kt (hunk 18 above)
Comment (juanii000) [reply app/src/main/java/com/example/parkingsystem_kotlin/mvp/view/ParkingView.kt]: Thanks

File: app/src/main/res/layout/dialog_fragment_parking_size.xml (hunk 19 above)
Comment (juanii000) [reply app/src/main/res/layout/dialog_fragment_parking_size.xml:28]: solved

**Review by mauroLonghi — APPROVED** [review]

**Review by nicolbidone — APPROVED** [review]
